share/eds/retriever.go

```diff
@@ -123,13 +128,24 @@ type retrievalSession struct {
 func (r *Retriever) newSession(ctx context.Context, dah *da.DataAvailabilityHeader) (*retrievalSession, error) {
 	size := len(dah.RowRoots)
 
+	adder := ipld.NewNmtNodeAdder(
+		ctx,
+		r.bServ,
+		ipld.MaxSizeBatchOption(size),
+	)
+
 	treeFn := func(_ rsmt2d.Axis, index uint) rsmt2d.Tree {
 		// use proofs adder if provided, to cache collected proofs while recomputing the eds
 		var opts []nmt.Option
-		visitor := ipld.ProofsAdderFromCtx(ctx).VisitFn()
-		if visitor != nil {
-			opts = append(opts, nmt.NodeVisitor(visitor))
+
+		proofsAdder := ipld.ProofsAdderFromCtx(ctx)
+		visitor := func(hash []byte, children ...[]byte) {
+			if proofsAdder != nil {
+				proofsAdder.VisitFn()(hash, children...)
+			}
+			adder.Visit(hash, children...)
```

Comment:
so in case of proofsAdder in ctx, both retrieval session's adder + proofs adder from ctx will cache collected proofs ?

Reply:
yes. Not a big deal, because soon, when shwap is released, we would not need proofsAdder here